docs: rewrite "Populating autocomplete with Sources" (#396)
algolia/autocomplete.js: packages/website/docs/sources.md

title: Populating autocomplete with Sources
---

> Sources define from where to retrieve items and their behavior.

Reviewer: Might be just me but I find this sentence with "from where" a bit hard to read.

Author: Yes, that's to avoid preposition stranding ("Sources define where to retrieve items *from*") which [some people really dislike](https://www.grammarly.com/blog/youve-been-lied-to-heres-why-you-absolutely-can-end-a-sentence-with-a-preposition/) because it's a bit too informal.

I'm fine with both as documentation isn't the most formal writing form. If that sounds better to you, I say we go for it.

Author: Handled in 5435db2.

        getItemUrl({ item }) {
          return item.url;
        },
        // ...
      },
    ];
  },
});
```

Before moving on to more complex sources, let's take a closer look at the code.

Notice that the [`getSources`](#getsources) function returns an array of sources. Each source implements a [`getItems`](#getitems) function to return the items to display. These items can be a simple static array, but you can also use a function to refine items based on the query. **The [`getItems`](#getitems) function is called whenever the input changes.**

> By default, autocomplete items are meant to be hyperlinks. To determine what URL to navigate to, you can implement a [`getItemURL`](#getitemurl) function. It enables the [keyboard accessibility](https://autocomplete.algolia.com/docs/keyboard-navigation) feature, allowing users to open items in the current tab, a new tab, or a new window from their keyboard.

Reviewer: ```suggestion
By default, autocomplete items are meant to be hyperlinks. To determine what URL to navigate to, you can implement a [`getItemURL`](#getitemurl) function. It enables the [keyboard accessibility](/docs/keyboard-navigation) feature, allowing users to open items in the current tab, a new tab, or a new window from their keyboard.
```

### Customizing items with templates

In addition to defining data sources for items, a source also lets you customize how to display items using [`templates`](#templates).

```js
autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          item({ item }) {
            return `Result: ${item.name}`;
          },
        },
      },
    ];
  },
});
```

> You can also display header and a footer elements around the list of items.

Reviewer: ```suggestion
You can also display header and footer elements around the list of items.
```

```js
autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          header() {
            return 'Suggestions';
          },
          item({ item }) {
            return `Result: ${item.name}`;
          },
          footer() {
            return 'Footer';
          },
        },
      },
    ];
  },
});
```

> **Templates aren't limited to strings.** You can provide anything as long as they're a valid virtual DOM element (see more in [**Templates**](templates)).

Reviewer: The link to **Templates** fails the build. Can we add the link in the other PR?

Author: Done in e0d1748.

});
```

:::note

The [`getItems`](#getitems) function supports [promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), meaning you can plug it to any asynchronous API.

:::

### Mixing static and dynamic sources based on the query

You don't have to show an empty screen until the user types a query. A typical pattern is to display a different source when the query is empty and switch once the user starts typing.

> ```js

Author: Addressed in 35d2550.

});
```

The [`getSources`](#getsources) function provides access to the current `query`, which you can use to return sources conditionally. You can use this pattern to display recent searches when the query is empty and search results when the user types.

> Note that you have access to the [complete autocomplete state](state), not only the query. It lets you compute sources based on many different aspects.

Reviewer: - "complete autocomplete" sound weird, don't you think?
- "aspects" is vague

```suggestion
Note that you have access to the [whole autocomplete state](state), not only the query. It lets you compute sources based on many different criteria.
```

Author: Right, I've switched with "full", and clarified the sentence with examples and a link to `state#state` in 9441f28.

### Using asynchronous sources

The [`getSources`](#getsources) function supports promises so that you can fetch sources from any asynchronous API. It can be Algolia or any third-party API you can query with an HTTP request.

For example, you could use the [Query Autocomplete](https://developers.google.com/places/web-service/query) service of the Google Places API to search for places and retrieve popular queries that map to actual points of interest.

```js
import { createAutocomplete } from '@algolia/autocomplete-core';

const autocomplete = createAutocomplete({
  getSources({ query }) {

>     return fetch(

Author: I've added an example using a third-party to showcase that Autocomplete isn't Algolia-exclusive.

The Google Places API and its Query Autocomplete service are an interesting example as they return both better queries and actual places, meaning we get a rather terse example where we can still showcase usage of `getItemInputValue`.

Reviewer: Great example—did you try it with an actual API key?

Author: No because you have to provide billing info to put your hands on a key. Do you have one yourself? Otherwise I can ask around.

Reviewer: You can try that one:

https://github.com/algolia/doc-code-samples/blob/ec699c32f572d946a705264e91243fa58229af4e/React%20InstantSearch/places/src/App.js#L38

Author: Already tried but it has a referrer restriction. It even doesn't work in Postman (bad request, not much help in the error message) with the example endpoint of their docs. My guess is that this API key doesn't have the  Places API enabled.

Reviewer: Ah true, I remember now that we restricted referrers for that API key.

Author: Let's grab a key later on when we want to build demos—for this one we can build a demo and take a video of it so that we don't expose our API key and get billed for tons of requests.

The code snippet should be fine, and worst case scenario, someone will PR with a fix.

});
```

:::note

You can use the official [`autocomplete-plugin-query-suggestions`](createQuerySuggestionsPlugin) plugin to retrieve Query Suggestions from Algolia.

:::

### Customizing items with templates

In addition to defining data sources for items, a source also lets you customize how to display items using [`templates`](#templates-specific-to-algoliaautocomplete-js).

:::info

Templates are only supported in [autocomplete-js](autocomplete-js).

> :::

Author: This is [what the document suggests later on](https://docs-1--algolia-autocomplete.netlify.app/docs/sources/#templates-specific-to-algoliaautocomplete-js). However, there's [another example](https://docs-1--algolia-autocomplete.netlify.app/docs/state#listening-to-state-changes) where `templates` is being used with `createAutocomplete`.

Is any of those outdated?

Reviewer: Good catch, that's an error.

I think we should default to documenting `@algolia/autocomplete-js` instead of `core`. It'll be less confusing and only power users will need `core`. Wdyt?

Author: So the error is that both `@algolia/autocomplete-core` and `@algolia/autocomplete-js` accept `templates`, right?

I agree that we should stick to `@algolia/autocomplete-js` examples as much as possible (not just for this case, but for the docs overall) unless we're specifically documenting `core`.

Reviewer: The templates concept only exists in `@algolia/autocomplete-js` so the error is to have used `createAutocomplete()` instead of `autocomplete()`.

Templates can return a string:

```js
import { autocomplete } from '@algolia/autocomplete-js';

autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          item({ item }) {
            return `Result: ${item.name}`;
          },
        },
      },
    ];
  },
});
```

Or a [Preact](https://preactjs.com/) component:

```js
import { autocomplete } from '@algolia/autocomplete-js';
import { h } from 'preact';

autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          item({ item }) {

>             return h('div', null, item.name);

Reviewer: ```suggestion
            return <div>{item.name}</div>;
            // This is equivalent to:
            // return h('div', null, item.name);
```

Author: Then we need to add a JSX pragma and we can drop the import of `h`, right?

Author: Handled in e9ea30d.

          },
        },
      },
    ];
  },
});

```

Or HTML/JSX-like syntax (here using [htm](https://github.com/developit/htm)):

```js
import { autocomplete, highlightHit } from '@algolia/autocomplete-js';
import { html } from 'htm/preact';

autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          item({ item }) {
            return html`<div>
              ${html([highlightHit({ hit: item, attribute: 'name' })])}
            </div>`
          },
        },
      },
    ];
  },
});
```

Or a custom component using `createElement` and `Fragment`:

```js
import { autocomplete } from '@algolia/autocomplete-js';

autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          item({ item, createElement, Fragment }) {
            return createElement(Fragment, {}, item.name);

>           },

Author: In [this sandbox](https://codesandbox.io/s/algoliajs-example-forked-298f6) I've found that `createElement` and `Fragment` are indeed exposed by `item`, however it doesn't seem that `root` is exposed.

According to the shape of [`AutocompleteSource`](https://github.com/algolia/autocomplete.js/blob/next/packages/autocomplete-js/src/types/AutocompleteSource.ts#L20-L24) there should be access to `root`, however I'm unsure where `createElement` and `Fragment` are coming from.

Reviewer: They're coming from #381 (not yet merged). We can quickly sync to make it clearer.

        },
      },
    ];
  },
});
```

You can also define `header` and `footer` templates to display before and after the items list.

```js
import { autocomplete } from '@algolia/autocomplete-js';

autocomplete({
  // ...
  getSources({ query }) {
    return [
      {
        // ...
        templates: {
          header() {
            return 'Suggestions';
          },
          // ...
          footer() {
            return 'Footer';
          },
        },
      },
    ];
  },
});
```

## Reference

### `getSources`

> `(params: { query: string, state: AutocompleteState, ...setters: Autocomplete Setters }) => Array<AutocompleteSource> | Promise<Array<AutocompleteSource>>`

The function to fetch sources and their behaviors.

A source implements the following interface:

> ### `getItems`

Reviewer: Theses should be level 4 (`####`) in terms of hierarchy, but they would then be hidden  by in the TOC by Docusaurus. Any recommendations?

Author: I would document the two sets separately [like here](https://deploy-preview-396--algolia-autocomplete.netlify.app/docs/state), one being "Sources" and the other "Source". This way we can clarify they aren't at the same level, yet keep them as level-3 headings.

Another way is to document them in the form `getSources[].getItems` or `getSources -> getItems` but I personally find it unsightly, idiosyncratic, and confusing.

> `(params: { state: AutocompleteState, ...setters, event: Event, item: TItem, itemInputValue: string, itemUrl: string, source: AutocompleteSource }) => void` | defaults to `({ setIsOpen }) => setIsOpen(false)`

Called whenever an item is selected.

### `onActive`

> `(params: { state: AutocompleteState, ...setters, event: Event, item: TItem, itemInputValue: string, itemUrl: string, source: AutocompleteSource }) => void`

Called whenever an item is active.

You can trigger different behaviors if the item is active depending on the triggering event using the `event` parameter.

### `templates`

> > `SourceTemplate`

Reviewer: ```suggestion
> `AutocompleteTemplate`
```

> > `AutocompleteTemplate`

Reviewer: ```suggestion
> `AutocompleteTemplates`
```

A set of templates to customize how items are displayed.

> You can also provide templates for header and a footer elements around the list of items.

Reviewer: ```suggestion
You can also provide templates for header and footer elements around the list of items.
```

Called to get the value of the item. The value is used to fill the search box.

> If you don't want to update the input value when an item is selected, you can return `state.query`.

Author: Isn't it easier not to implement the function?

Reviewer: Yes, especially that we document the default value.

Author: Dropped in 62cc97b.

> `(params: { state: AutocompleteState, ...setters, event: Event, item: TItem, itemInputValue: string, itemUrl: string, source: AutocompleteSource }) => void` | defaults to `({ setIsOpen }) => setIsOpen(false)`

Called whenever an item is selected.

### `onActive`

> `(params: { state: AutocompleteState, ...setters, event: Event, item: TItem, itemInputValue: string, itemUrl: string, source: AutocompleteSource }) => void`

Called whenever an item is active.

You can trigger different behaviors if the item is active depending on the triggering event using the `event` parameter.

### `templates` (specific to `@algolia/autocomplete-js`)

> `SourceTemplate`

The `@algolia/autocomplete-js` supports source templates.

> A template can either return a string or perform DOM mutations without returning a string.

Author: This is the sentence from the [source code](https://github.com/algolia/autocomplete.js/blob/next/packages/autocomplete-js/src/types/AutocompleteSource.ts#L14) but I find it confusing.

What do you mean by "perform DOM mutations without returning a string"? If you don't return anything, the template will be empty, won't it?

Reviewer: This is not true anymore from #381 (which will get merged soon once coordinated with other open PRs).

As documented above, we now support strings, components, etc. so you can ignore this outdated formulation.

Author: Handled here 2ce1218.
